File: plugins/OrchidForms/src/main/java/com/eden/orchid/presentations/FormOptionExtractor.java
package com.eden.orchid.presentations;

import com.eden.common.util.EdenPair;
import com.eden.orchid.api.OrchidContext;
import com.eden.orchid.api.converters.StringConverter;
import com.eden.orchid.api.options.OptionExtractor;
import com.google.inject.Provider;
import org.json.JSONObject;

import javax.inject.Inject;
import java.lang.reflect.Field;
import java.util.List;

public final class FormOptionExtractor extends OptionExtractor<Form> {

    private final Provider<OrchidContext> contextProvider;
    private final StringConverter converter;
    private final FormsModel formsModel;

    @Inject
    public FormOptionExtractor(Provider<OrchidContext> contextProvider, StringConverter converter, FormsModel formsModel) {
        super(1000);
        this.contextProvider = contextProvider;
        this.converter = converter;
        this.formsModel = formsModel;
    }

    @Override
    public boolean acceptsClass(Class clazz) {
        return clazz.equals(Form.class);
    }

    @Override
    public Form getOption(Field field, JSONObject options, String key) {
        if(options.has(key)) {
            if(options.get(key) instanceof JSONObject) {
                return new Form(contextProvider.get(), key, options.getJSONObject(key));
            }
            else {
                EdenPair<Boolean, String> value = converter.convert(options.get(key));
                if(value.first) {
                    return formsModel.getForms().getOrDefault(value.second, null);
                }
            }
        }

        return getDefaultValue(field);
    }

    @Override
    public Form getDefaultValue(Field field) {
        return null;
    }

    @Override
    public List<Form> getList(Field field, JSONObject options, String key) {
        return null;
    }

    @Override
    public Object[] getArray(Field field, JSONObject options, String key) {
        return null;
    }
}

File: plugins/OrchidForms/src/main/kotlin/com/eden/orchid/forms/FormOptionExtractor.kt
package com.eden.orchid.forms

import com.eden.orchid.api.OrchidContext
import com.eden.orchid.api.converters.StringConverter
import com.eden.orchid.api.options.OptionExtractor
import com.eden.orchid.forms.model.Form
import com.eden.orchid.forms.model.FormsModel
import com.eden.orchid.utilities.SuppressedWarnings
import com.eden.orchid.utilities.resolve
import javax.inject.Provider
import org.json.JSONObject
import java.lang.reflect.Field
import javax.inject.Inject

class FormOptionExtractor
@Inject
constructor(
        private val contextProvider: Provider<OrchidContext>,
        private val converter: StringConverter
) : OptionExtractor<Form>(1000) {

    override fun acceptsClass(clazz: Class<*>): Boolean {
        return clazz == Form::class.java
    }

    @Suppress(SuppressedWarnings.UNCHECKED_KOTLIN)
    override fun getOption(field: Field, sourceObject: Any, key: String): Form? {
        val formsModel: FormsModel = contextProvider.get().resolve()
        if (sourceObject is JSONObject) {
            return Form(contextProvider.get(), "", sourceObject.toMap())
        }
        else if (sourceObject is Map<*, *>) {
            return Form(contextProvider.get(), "", sourceObject as? Map<String, Any> ?: emptyMap())
        }
        else {
            val value = converter.convert(String::class.java, sourceObject)
            if (value.first) {
                return formsModel.forms.getOrDefault(value.second, null)
            }
        }

        return null
    }

    override fun getDefaultValue(field: Field): Form? {
        return null
    }

}
